src/llmq/signing_shares.cpp

```diff
@@ -1481,36 +1492,76 @@ void CSigSharesManager::BanNode(NodeId nodeId, PeerManager& peerman)
     });
     nodeState.requestedSigShares.Clear();
     nodeState.banned = true;
+    // Banning affects request routing and can create immediate work
+    NotifyWorker();
 }
 
 void CSigSharesManager::WorkThreadMain(CConnman& connman, PeerManager& peerman)
 {
-    int64_t lastSendTime = 0;
-
     while (!workInterrupt) {
+        // capture epoch at loop start to detect intervening notifications
+        uint64_t startEpoch = workEpoch.load(std::memory_order_acquire);
         RemoveBannedNodeStates(peerman);
 
         bool fMoreWork = ProcessPendingSigShares(peerman, connman);
         SignPendingSigShares(connman, peerman);
 
-        if (TicksSinceEpoch<std::chrono::milliseconds>(SystemClock::now()) - lastSendTime > 100) {
-            SendMessages(connman);
-            lastSendTime = TicksSinceEpoch<std::chrono::milliseconds>(SystemClock::now());
-        }
+        bool didSend = SendMessages(connman);
 
         Cleanup(connman);
 
-        // TODO Wakeup when pending signing is needed?
-        if (!fMoreWork && !workInterrupt.sleep_for(std::chrono::milliseconds(100))) {
-            return;
+        // If there is more work or we just sent something, iterate again without waiting
+        if (fMoreWork || didSend) {
+            continue;
+        }
+
+        // Compute next wake-up deadline for periodic tasks (recovery attempts and cleanup cadence)
+        auto next_deadline = std::chrono::steady_clock::time_point::max();
+        // Respect cleanup cadence (~5s) even when idle
+        {
+            auto now_tp = std::chrono::steady_clock::now();
+            int64_t now_s = GetTime<std::chrono::seconds>().count();
+            int64_t target_s = lastCleanupTime + 5;
+            if (target_s > now_s) {
+                auto delta_ms = (target_s - now_s) * 1000;
+                auto cand = now_tp + std::chrono::milliseconds(delta_ms);
+                if (cand < next_deadline) next_deadline = cand;
```

Comment:
This logic is broken imo - you can't set system time point based on the difference of mocked ones (`delta_ms`), mocked time moves independently